src/string/shortest_palindrome.rs

```diff
@@ -0,0 +1,73 @@
+/*
+The function shortest_palindrome expands the given string to shortest palindrome by adding a shortest prefix.
+KMP. Source：https://www.scaler.com/topics/data-structures/kmp-algorithm/
+Prefix Functions and KPM. Source：https://oi-wiki.org/string/kmp/
+*/
+
+pub fn shortest_palindrome(s: &str) -> String {
+    if s.is_empty() {
+        return "".to_string();
+    }
+
+    let p_chars: Vec<char> = s.chars().collect();
+    let suffix = raw_suffix_function(s, p_chars.clone());
+
+    let mut s_chars: Vec<char> = s.chars().rev().collect();
+    // The prefix of the original string matches the suffix of the flipped string.
+    let dp = invert_suffix_function(s, p_chars.clone(), s_chars.clone(), suffix);
+
+    s_chars.append(&mut p_chars[dp[s.chars().count() - 1]..s.chars().count()].to_vec());
```

Comment:
computing `s.chars().count()` is very expensive. The number of characters can be obtained from `p_chars`.